numba/tests/test_lists.py

```diff
@@ -865,5 +865,266 @@ def test_reflect_clean(self):
+        def pyfunc():
+            con = []
+            for i in range(5):
+                con.append(np.arange(2))
+            return con[2:4]
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_set_slice(self):
+        def pyfunc():
+            con = []
+            for i in range(5):
+                con.append(np.arange(2))
+            con[1:3] = con[2:4]
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_pop(self):
+        def pyfunc():
+            con = []
+            for i in range(20):
+                con.append(np.arange(i + 1))
+            while len(con) > 2:
+                con.pop()
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_pop_loc(self):
+        def pyfunc():
+            con = []
+            for i in range(1000):
+                con.append(np.arange(i + 1))
+            while len(con) > 2:
+                con.pop(1)
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_del_range(self):
+        def pyfunc():
+            con = []
+            for i in range(20):
+                con.append(np.arange(i + 1))
+            del con[3:10]
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_list_of_list(self):
+        def pyfunc():
+            con = []
+            for i in range(10):
+                con.append([0] * i)
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assertEqual(expect, got)
+
+    def test_list_of_list_reflected(self):
+        def pyfunc(l1, l2):
+            l1.append(l2)
+            l1[-1].append(123)
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l1 = [[0, 1], [2, 3]]
+        l2 = [4, 5]
+        expect = list(l1), list(l2)
+        got = list(l1), list(l2)
+        pyfunc(*expect)
+        cfunc(*got)
+        self.assertEqual(expect, got)
+
+    def test_heterogenous_list(self):
+        def pyfunc(x):
+            return x[1]
+
+        l1 = [[np.zeros(i) for i in range(5)], [np.ones(i) for i in range(5)]]
+        l2 = [[np.zeros(i) for i in range(5)], [np.ones(i)+1j for i in range(5)]]
```

Comment:
@stuartarchibald , For heterogeneous lists that have a promotion path, it will work.  Here the first element is of `array<double>` and the second element is of `array<complex128>`.  The double dtype is promoted to complex128.  The promotion rule is true for the scalar version when list couldn't contain arrays.  I guess it is unexpected for anything that are not numbers.

Reply:
Thanks, I think this has to be the case to support more use cases.